DatadogRUM/Sources/Instrumentation/AppHangs/AppHangsWatchdogThread.swift

```diff
@@ -0,0 +1,131 @@
+        ///
+        /// Introducing even a small sleep significantly reduces CPU consumption by the watchdog thread (nearly 0%). No sleep would result in close to 100% CPU usage.
+        ///
+        /// `tolerance` defines the margin of error in hang monitoring. The "measured hang duration" may differ from `0` to `tolerance` % when compared to the "actual hang duration".
+        /// This means some actual hangs very close to `appHangThreshold` might be considered false negatives and not reported.
+        ///
+        /// **Example:**
+        /// If `appHangThreshold` is 2 seconds and tolerance is 2.5%, the watchdog thread will idle for 50 milliseconds before taking each sample of the main thread.
+        /// Because the sleep might start right at the moment of hang beginning on the main thread, the watchdog thread will miss the first 50 milliseconds of hang duration.
+        /// If the actual hang lasts exactly 2 seconds, the watchdog will measure it as 1950 milliseconds. As a result, the SDK will not report it as it falls under `appHangThreshold`.
+        static let tolerance: Double = 0.025 // 2.5%
+    }
+
+    /// The minimal duration of the main thread hang to consider it an App Hang.
+    private let appHangThreshold: TimeInterval
+    /// The "idle" interval for watchdog thread operations. Reduces the pressure on CPU.
+    private let idleInterval: TimeInterval
+    /// Semaphore used to block this thread until main thread responds.
+    private let mainThreadTask = DispatchSemaphore(value: 0)
+    /// The queue to observe by this thread (main queue).
+    private let mainQueue: DispatchQueue
+    /// SDK date provider.
+    private let dateProvider: DateProvider
+    /// Telemetry interface.
+    private let telemetry: Telemetry
+    /// Closure to be notified when App Hang ends. It will be executed on the watchdog thread.
+    @ReadWriteLock
+    internal var onHangEnded: ((AppHang) -> Void)?
+
+    /// Creates an instance of an App Hang watchdog thread.
+    ///
+    /// This thread is not started by default and requires manual invocation of `.start()`.
+    ///
+    /// - Parameters:
+    ///   - appHangThreshold: Minimum duration of the `queue` hang to consider it an App Hang.
+    ///   - queue: The queue to observe for hangs. (main queue)
+    ///   - dateProvider: Date provider.
+    ///   - telemetry: The handler to report issues through RUM Telemetry.
+    init(
+        appHangThreshold: TimeInterval,
+        queue: DispatchQueue,
+        dateProvider: DateProvider,
+        telemetry: Telemetry
+    ) {
+        self.appHangThreshold = appHangThreshold
+        self.idleInterval = appHangThreshold * Constants.tolerance
+        self.mainQueue = queue
+        self.dateProvider = dateProvider
+        self.telemetry = telemetry
+        super.init()
+        self.name = "com.datadoghq.app-hang-watchdog"
+    }
+
+    override func main() {
+        let mainThreadTask = self.mainThreadTask
```

Comment:
what's the benefit of doing that here?

Reply:
I tried to avoid capturing `[weak self]` as part of the frequently scheduled task on `mainQueue`. There's no gain, just practical choice (semaphore is a class, so this is correct).